compiler/rustc_typeck/src/check/method/probe.rs

```diff
@@ -1143,6 +1160,33 @@ impl<'a, 'tcx> ProbeContext<'a, 'tcx> {
         })
     }
 
+    /// If `self_ty` is `*mut T` then this picks `*const T` methods. The reason why we have a
+    /// special case for this is because going from `*mut T` to `*const T` with autoderefs and
+    /// autorefs would require dereferencing the pointer, which is not safe.
```

Comment:
IIUC, it's not so much that the pointer would be dereferenced, but more that adding an autoref would create a safe reference, which we don't want to do.

Reply:
I just tried this:

```rust
struct Test {}

impl Test {
    fn x(&self) {
        println!("x called");
    }
}

fn main() {
    let mut test = Test {};
    let mut test_mut_ref = &mut test;
    test_mut_ref.x();
}
```

the relevant, unoptimized MIR:

```
        _4 = &(*_2);
        _3 = Test::x(move _4) -> [return: bb1, unwind: bb2];
```

I thought the `*` dereferences, no? Similar code for the pointer version would also do `&*` and dereference.

Reply:
You're not wrong, the `*` does dereference -- it's just that, when you combine it with a `&`, the net effect is kind of an "identity transform", at least at runtime. That is, `&*x` is sort of equivalent to `x` (the "address of the data that `x` points to" is kind of just "x") -- but only *sort of*, because now we've made a shared reference, and that has implications (e.g., if `x: &mut T`, it implies that `*x` is frozen while this shared reference is live).

Reply:
Is this the case even when result of `&` is a pointer, rather than reference? I'm guessing the final assembly for `&*` (for pointer-to-pointer conversions) may disappear, but I thought in MIR it's still dereferencing, and has the implications of `*` operation (whatever those implications are).

Instead what we do here is more like `x as *const T` (where `x : *mut T`), which doesn't have unsafe dereferencing in MIR, and as a result would not have the same implications.